cookbooks/equilibrium_grain_size/plugins/equilibrium_grain_size.cc

```diff
@@ -0,0 +1,1750 @@
+
+                      out.densities[i] = craton_density * in.composition[i][craton_index] +
+                                         out.densities[i] * (1. - in.composition[i][craton_index]);
+                    }
+                }
+              else if (depth > lithosphere_thickness && depth <= uppermost_mantle_thickness)
+                {
+                  out.thermal_expansion_coefficients[i] = 3.e-5;
+                  out.compressibilities[i] = 1./12.2e10;
+                  out.densities[i] = 3.3e3 * (1. - out.thermal_expansion_coefficients[i] * deltaT
+                                              + pressure * out.compressibilities[i]);
+                  material_type = 3;
+                }
+              else
+                {
+                  // Densities below 300 km are computed using the scaling relationship from the velocity anomalies
+                  double density_vs_scaling;
+                  if (use_depth_dependent_rho_vs)
+                    density_vs_scaling = rho_vs_depth_profile.get_data_component(Point<1>(depth), density_scaling_index);
+                  else
+                    // Values from Becker [2006], GJI
+                    density_vs_scaling = 0.15;
+
+                  const double density_anomaly = delta_log_vs * density_vs_scaling;
+                  const double reference_density = this->get_adiabatic_conditions().is_initialized()
+                                                   ?
+                                                   this->get_adiabatic_conditions().density(in.position[i])
+                                                   :
+                                                   reference_rho;
+
+                  out.densities[i] = reference_density * (1. + density_anomaly);
+
+                  if (use_depth_dependent_thermal_expansivity)
+                    out.thermal_expansion_coefficients[i] = thermal_expansivity_profile.get_data_component(Point<1>(depth), thermal_expansivity_column_index);
+                  else
+                    out.thermal_expansion_coefficients[i] = thermal_alpha;
+
+                  out.compressibilities[i] = reference_compressibility;
+
+                  material_type = 4;
+                }
+
+              if (MaterialTypeAdditionalOutputs<dim> *material_type_out = out.template get_additional_output<MaterialTypeAdditionalOutputs<dim>>())
+                material_type_out->output_values[0][i] = material_type;
+            }
+        }
+    }
+
+
+
+    template <int dim>
+    void
+    EquilibriumGrainSize<dim>::declare_parameters (ParameterHandler &prm)
+    {
+      prm.enter_subsection("Material model");
+      {
+        prm.enter_subsection("Equilibrium grain size model");
+        {
+          prm.declare_entry ("Reference density", "3300",
+                             Patterns::Double (0),
+                             "The reference density $\\rho_0$. Units: $kg/m^3$.");
```

Comment:
Here and for all of the parameters below: We try to use semantic mark-up for units. For example, this could look like this:
```
Units: \\si{\\meter}.
```
Take a look at some of the other files